capstone/backend/src/main/java/com/google/servlets/StoreRankingsServlet.java

```diff
@@ -0,0 +1,179 @@
+        private String status;
+
+        private class Distance {
+          String text;
+          String value;
+        }
+
+        private class Duration {
+          String text;
+          String value;
+        }
+      }
+    }
+
+    private String status;
+    private List<String> origin_addresses;
+    private List<String> destination_addresses;
+    private List<Row> rows;
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserPreferences userPreferences = getUserPreferencesAndValidate(request);
+    if (userPreferences == null) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid request syntax.");
+      return;
+    }
+    List<Store> stores = new ArrayList<Store>();
+    try {
+      stores = getStores(userPreferences.getSelectedItemTypes());
+    } catch (SpannerException e) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, e.getMessage());
+      return;
+    }
+    Map<String, Integer> distances =
+        getDistances(
+            stores.stream().map(store -> store.getStoreAddress()).collect(Collectors.toList()),
+            userPreferences.getLocation());
+    stores =
+        stores.stream()
+            .filter(
+                store ->
+                    (distances.get(store.getStoreAddress())
+                        < (userPreferences.getDistancePreference() * MILES_TO_METERS)))
+            .collect(Collectors.toList());
+    rankStores(stores, distances);
+    response.setContentType("application/json");
+    response.setStatus(HttpServletResponse.SC_OK);
+    response.getWriter().println(g.toJson(stores));
+  }
+
+  private void rankStores(List<Store> stores, Map<String, Integer> distances) {
+    Collections.sort(
+        stores,
+        new Comparator<Store>() {
+          @Override
+          public int compare(Store s1, Store s2) {
+            double s1StoreScore =
+                s1.getNumberOfItemsFound() * AVALIABLE_ITEMS_WEIGHT
+                    + s1.getLowestPotentialPrice() * PRICE_WEIGHT
+                    + distances.get(s1.getStoreAddress()) * DISTANCE_WEIGHT;
+            double s2StoreScore =
+                s2.getNumberOfItemsFound() * AVALIABLE_ITEMS_WEIGHT
+                    + s2.getLowestPotentialPrice() * PRICE_WEIGHT
+                    + distances.get(s2.getStoreAddress()) * DISTANCE_WEIGHT;
+            if (s1StoreScore < s2StoreScore) {
+              return -1;
+            } else if (s1StoreScore > s2StoreScore) {
+              return 1;
+            } else {
+              return 0;
+            }
+          }
+        });
+  }
+
+  public Map<String, Integer> getDistances(List<String> addresses, List<Double> userLocation)
+      throws IOException {
+    HttpRequestFactory requestFactory = new NetHttpTransport().createRequestFactory();
+    StringBuilder sb =
+        new StringBuilder("https://maps.googleapis.com/maps/api/distancematrix/json?origins=");
```

Comment:
Can try using a URL encoder/URL builder library here, then you don't have to do the parsing manually. Something like
https://www.baeldung.com/java-url-encoding-decoding